src/clj/fluree/db/flake/reasoner.cljc

```diff
@@ -163,28 +165,52 @@
     (log/debug "Reasoner - source OWL rules: " graph)
     (owl-datalog/owl->datalog inserts graph)))
 
+(defn rules-from-dbs
+  [methods inserts dbs]
+  (let [clause-chan (async/to-chan! dbs)
+        out-chan (async/chan)]
+    (doall
+     (for [method methods]
+       (async/pipeline-async
+        1
+        out-chan
+        (fn [db out-ch]
+          (async/pipe (go-try
```

Comment:
I'm trying to loop over the `clause-chan` with something like this:
```
(defn rules-from-dbs
  [methods inserts dbs]
  (let [clause-chan (async/to-chan! dbs)
        out-chan (async/chan)]
    (doall
     (for [method methods]
       (clojure.core.async/go-loop [from-ch clause-chan]
         (let [db (<? from-ch)]
           (if (nil? db)
             out-chan
             (do (async/pipe (go-try
                               (as-> db $
                                 (<? (resolve/rules-from-db $ method))
                                 (rules-from-graph method inserts $)))
                             out-chan)
                 (recur from-ch)))))))))
```
Maybe I misunderstood what you were saying, but I'm just replacing the `pipeline-async` call with a `go-loop`. With `go-loop` I can process elements from the `from-chan` and `pipe` them onto the `out-chan`, but this ends up requiring me to call `<?` twice on the results since `go-loop` creates a channel that contains the `out-chan`. So I think I'm doing something wrong, but I'm not sure what.

Reply:
`<?` throws any exceptions that come off of a channel immediately. When an exception is thrown inside of a go block, it is silently swallowed and the channel returned by the go block closes without returning any more results. Therefore, you need to explicitly handle any errors that could possibly be thrown with a `try`/`catch` block or something similar. 

You should only use the higher order pipeline functions when you are certain that none of the code inside the pipeline can throw for this reason, or you are certain that you have handled any possible errors that could be thrown.

You need to account for the possiblity that `<?` is an exception, and do the right thing in each case it is. If it's impossible for it to be an exception, then you shouldn't use `<?`; you should use `<!` instead.

Reply:
This sounds like it is all assuming that I would be going the route of using the higher level pipeline functions and explicitly handling any exceptions. You were originally recommending that I use a loop though, which is what I was struggling with figuring out. Is there one of these solutions that is better than the other?

Reply:
Right, but your first comment on this thread was about `<?` not working with the `pipeline-` functions, but now it sounds like you are just talking about `<?` needing explicit error handling regardless of whether I'm using the `pipeline-` functions. So should I just keep the `pipeline-` functions and do explicit error handling since it sounds like I will need to do that either way.

Reply:
Here's what I said:
> go-try/<? does not work well with the higher order core.async pipeline functions like pipeline-async and pipe because any errors that are thrown won't be handled properly and instead be propagated downstream.

`go-try` takes any exception that is thrown inside it's block and puts it on its output channel. `<?` takes any exception from the channel it's pulling from and throws it. That behavior is not compatible with the pipeline functions because the whole point of using them is so you can process collections as a single unit. That no longer works when there could be exception objects mixed in to those collections. 

If you are going to use the pipeline functions, then you need to make sure that none of the code in the pipeline functions can throw exceptions (which `<?` does), and you need to make sure that none of the code upstream of those functions will put exceptions on their input channel, which `go-try` does.